### DISCUSSION TITLE
Force the usage of embedded clang-format

### DISCUSSION THREAD

**Discussion by KaeLL** [microsoft/vscode-cpptools#7117] [Q&A] [answered]
Is there a way to do it? Seems like it's intended to be a last-resort fallback, but I'd like to use it as a first class option, because of the redundancy of having more than one on my machine, which is pointless since I'm always working on VSCode, and even if that wasn't the case, it's generally more up-to-date than my distro's.
Also, can it be used interactively? As in, is there a way to communicate with it, sort of like we do with the regular clang-format? I wonder if it can completely replace my distro's clang-format.

**Comment by sean-mcmanus** [comment, marked as answer]
The bundled version of clang-format is supposed to automatically be used if the version is newer. Is that not working for you? Are you sure you don't have the C_Cpp.clang_format_path setting set? The code that does this is around https://github.com/microsoft/vscode-cpptools/blob/main/Extension/src/LanguageServer/settings.ts#L106 if you feel like debugging what is going wrong.

**Reply by KaeLL** [reply]
> The bundled version of clang-format is supposed to automatically be used if the version is newer.

TIL.
> Is that not working for you?

Probably yes.
> Are you sure you don't have the C_Cpp.clang_format_path setting set?

I'm sure.

> The code that does this is around https://github.com/microsoft/vscode-cpptools/blob/main/Extension/src/LanguageServer/settings.ts#L106 if you feel like debugging what is going wrong.

Now I found out where you store your clang-format. Thanks.

**Reply by sean-mcmanus** [reply]
The TypeScript code appears to be working (see the screenshot), but it's possible there might be an issue in the cpptools process? How do you know the correct clang-format is not being used? What versions are you using?
![image](https://user-images.githubusercontent.com/19859882/110185769-1e498d00-7dc8-11eb-9d79-0e84779bea52.png)

**Reply by sean-mcmanus** [reply]
If you set
```
    "C_Cpp.formatting": "clangFormat",
    "C_Cpp.clang_format_style": "{ BasedOnStyle: GNU }",
    "C_Cpp.loggingLevel": "Debug"
```
You should see output like:
```
Formatting Engine: clangFormat
Executing command (Mac/Linux) (wait_for_exit=true): /home/user/vscode-cpptools/Extension/bin/../LLVM/bin/clang-format
  Arguments:
    -style={ BasedOnStyle: GNU }
 ```
i.e. I'm not detecting any bug (the GNU style got added with clang-format 11, which is what is bundled with 1.2.2). Let me know if you have more repro info.

**Reply by KaeLL** [reply]
I tested it, and even though it didn't show
```
Executing command (Mac/Linux) (wait_for_exit=true): /home/user/vscode-cpptools/Extension/bin/../LLVM/bin/clang-format
  Arguments:
    -style={ BasedOnStyle: GNU }
```
it worked correctly.

**Reply by sean-mcmanus** [reply]
That means it's using the bundled clang-format 11.0 -- is your issue resolved or did you think there was a bug still?

**Reply by KaeLL** [reply]
There was no bug. There's also no trace of all but the 1st line of the log you mentioned like I said in my previous comment. Could that be a bug?! The way I tested the bundled clang-format was being used is by setting the GNU style like you said and formatting code, which worked fine, so I assumed it.

**Reply by sean-mcmanus** [reply]
I was just confused -- the "Executing command" logging only appears with our unreleased Debug bits.
